revolution: use seppuku memmap if ext flash absent

mlyle opened the pull request.

mlyle committed:
revolution: use seppuku memmap if ext flash absent

mlyle commented:
Untested.

jenkins, build artifacts please

mlyle commented:
whoops, first cut wasn't quite right.

jenkins, build artifacts please

mlyle commented:
jenkins, test this please, too, because i cancelled the wrong build :<

LitterBugs commented:
Tested on Sunrise Model Cicada F4+BLHeli_S4in1.  No bootloader found via rescue.  No lights.

mlyle committed:
revolution: move settings sector size to 0x20000

This means on external flash we have effectively 3 places to wear
level over, storing twice the settings (and writing less than half
as often for the same amount of settings)

mlyle commented:
@litterbugs -- OK, just to confirm-- you flashed a new EF from this image?  I found a problem that would prevent the firmware from starting but not the BL.   (Will inspect a little more).

LitterBugs commented:
@mlyle  Correct.  I flashed from the image posted in this thread.  No bootloader or firmware response.

mlyle commented:
OK.  I'm a bit confused, but thought of a couple of things to try.

jenkins, build artifacts please

P.S. the sparky2 bl running on it is valuable info, but I don't recommend it-- the pinmaps aren't the same and I can't promise it won't damage hardware.

LitterBugs commented:
Understood about incorrect firmware for board.  Only powered  briefly over USB for testing purposes.

mlyle commented:
Ugh, had to go kick the Windows builder. :P

LitterBugs commented:
Flashed latest revolution artifacts .  No rescue response.  No light.

Flashed back to BetaFlight  OmnibusF4 and works fine in configurator.

LitterBugs commented:
from BF CLI:

# status
System Uptime: 264 seconds
Voltage: 0 * 0.1V (0S battery - NOT PRESENT)
CPU Clock=168MHz, GYRO=MPU6000, ACC=MPU6000
Stack size: 2048, Stack address: 0x10010000
I2C Errors: 0, config size: 2078, max available config: 16384
CPU:4%, cycle time: 128, GYRO rate: 7812, RX rate: 49, System rate: 9
Arming disable flags: ANGLE CLI


# resource
resource BEEPER 1 B04
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A03
resource MOTOR 4 A02
resource MOTOR 5 A01
resource MOTOR 6 A08
resource PPM 1 B14
resource PWM 1 B14
resource PWM 2 B15
resource PWM 3 C06
resource PWM 4 C07
resource PWM 5 C08
resource PWM 6 C09
resource LED_STRIP 1 A01
resource SERIAL_TX 1 A09
resource SERIAL_TX 3 B10
resource SERIAL_TX 6 C06
resource SERIAL_RX 1 A10
resource SERIAL_RX 3 B11
resource SERIAL_RX 6 C07
resource INVERTER 1 C00
resource LED 1 B05
resource SPI_SCK 1 A05
resource SPI_SCK 3 C10
resource SPI_MISO 1 A06
resource SPI_MISO 3 C11
resource SPI_MOSI 1 A07
resource SPI_MOSI 3 C12
resource ESCSERIAL 1 B14
resource ADC_BATT 1 C02
resource ADC_RSSI 1 A00
resource ADC_CURR 1 C01

mlyle commented:
Aha, found a typo.

jenkins, build artifacts please

LitterBugs commented:
Really close....  Getting the usual "._." R  flash from the LED now.  It shows up as USB Revolution in the GCS connections box.  But GCS can not communicate with it and Rescue does not find it either.  Tried flashing a second time, but still can't communicate.

mlyle commented:
@LitterBugs ah.  That means the USB sense line isn't connected.  I had had something in before for that but took it out when it didn't seem to be the main problem ;)

Hmm.  Let me think about how to handle that properly-- not having usb vsense breaks DSM binding, so I want to make sure to only break it for the non-eeprom variants.

mlyle committed:
pios_usb: new call tells if vsense line present

mlyle committed:
flight: autodetermine bootloader pause delay

Cuts down on the amount of per-target code, and also allows different
variants of targets to have vsense or not.

mlyle committed:
revolution: if extflash missing, assume no vsense

People who halfass one seem to halfass the other :P

mlyle commented:
jenkins, build artifacts please

mlyle commented:
OK I'm moderately hopeful with this, though the change keeps getting ever-more-complicated.

BetaFlight doesn't care whether the USB cable is plugged in and tries to communicate either way-- we don't.  Now if the EEPROM is missing we'll ignore the USB plugged-in status.

mlyle commented:
Big remaining concern is this:

```
#if defined(OMNIBUSF4SD)
    DEF_TIM(TIM4,  CH3, PB8,  TIM_USE_PWM | TIM_USE_PPM,   TIMER_OUTPUT_NONE,     0), // PPM
    DEF_TIM(TIM4,  CH4, PB9,  TIM_USE_PWM,                 TIMER_OUTPUT_NONE,     0), // S2_IN
#else
    DEF_TIM(TIM12, CH1, PB14, TIM_USE_PWM | TIM_USE_PPM,   TIMER_OUTPUT_NONE,     0), // PPM
    DEF_TIM(TIM12, CH2, PB15, TIM_USE_PWM,                 TIMER_OUTPUT_NONE,     0), // S2_IN
#endif
```

There's two OmnibusF4 variants.  The non-SD put PWM1&2 on the same pins as revolution, the SD variants put them on different pins.

mlyle commented:
Sorry, no, misread.  This is input pins, for what pin is used for PPM.  If you use SBUS etc it should not be affected.

LitterBugs commented:
DING DING DING!   We HAVE a WINNER!   Guess now I have to build something with it to give it a full test.  Just been playing over USB up to this point.   Looks like the sensor is yaw rotated 90 degrees.  Think I can handle that.

mlyle commented:
Nice, thanks for your patience and trying all these :D   

Unfortunately this will take some effort to merge and prove-- to make sure it doesn't mess up normal Revolution and that the effects on other targets aren't bad.

LitterBugs commented:
No worries!  You went above and beyond to get it working.  I had just planned on settling for BF on it when it didn't work.  Will still use BF on it for 3D mode.  Been wanting to try that for a long time.

mlyle committed:
flight: pios_annunc: cope with missing annunciators

mlyle committed:
revolution: properly handle annunciator on OmniF4

mlyle committed:
revolution: bootloader: bump version

mlyle committed:
revolution: support omnibus f4 v2 LED pin

mlyle commented:
Requires test on real revolution, including bootloader, to merge.  I will try and do this Weds, speedy merge afterwards appreciated.

mlyle commented:
rebased

mlyle commented:
Confirmed: OK on revolution, early clones.

I witnesssed uploader doing weird things a couple of times, freezing etc -- which it's worth keeping an eye out for to make sure that there's not a new-qt related regression.  But that doesn't relate to this patch set.

tracernz commented on flight/PiOS/inc/pios_usb.h:
```diff
@@ -36,6 +36,7 @@ extern int32_t PIOS_USB_Reenumerate();
 extern int32_t PIOS_USB_ChangeConnectionState(bool connected);
 extern bool PIOS_USB_CableConnected(uintptr_t id);
 extern bool PIOS_USB_CheckAvailable(uintptr_t id);
+extern bool PIOS_USB_HaveVsense(uintptr_t id);
```
s/s/S/

mlyle replied:
Fixed in latest commit

tracernz commented on flight/targets/revolution/bl/pios_board.c:
```diff
@@ -72,11 +72,13 @@ void PIOS_Board_Init()
 	PIOS_Flash_Internal_Init(&pios_internal_flash_id, &flash_internal_cfg);
 
+	if (!flash_chip_ok) {
+		pios_external_flash_id = 0;
```
I would say `NULL`, but this is still an int? :(

mlyle replied:
Yep

tracernz commented on flight/targets/revolution/fw/pios_board.c:
```diff
@@ -114,8 +114,12 @@ void PIOS_Board_Init(void) {
+		/* Cope with clones with no flash, too. */
+
+		is_modified_clone = true;
+		pios_external_flash_id = 0;
```
This is a real shame for proper revo. :disappointed:

tracernz commented on flight/targets/revolution/board-info/pios_board.h:
```diff
@@ -86,10 +86,11 @@ TIM8  |           |           |           |
 //------------------------
 #define PIOS_LED_HEARTBEAT	0
 #define PIOS_LED_ALARM		1
+#define PIOS_ANNUNCIATOR_BUZZER	2
```
Alignment broken by this

mlyle replied:
Not in reality just in diff (extra + at the beginning throws it off)

tracernz replied:
Oh yeah. Should really use spaces for aligning stuff anyway (but still tabs for indentation ofc).

mlyle committed:
pios_usb: Have_Vsense -> Have_VSense
